### DISCUSSION TITLE
Question: how to deal with object mapping, domain entities construction?

### DISCUSSION THREAD

**Discussion by MaKTD** [mikro-orm/mikro-orm#1380] [Q&A] [answered]
Hello, first of all, thank you for your product, the unit of work and identity map really shines. However, I didn't find some answers in docs and issues. My question generally about how to plug in micro-orm to the project which was designed highly in ddd.

Let's imagine that project uses value objects, entities. For example we have can have email value object which will look like this.
```
export class Email extends ValueObject<EmailProps> {
  get value(): string {
    return this.props.value
  }

  private constructor(props: EmailProps) {
    super(props)
  }

  public static create(email: string): Result<Email> {
    const guardResult = Guard.usingJSONSchema(
      { type: 'string', format: 'email' },
      email,
    )
    if (!guardResult.succeeded) return Result.fail<Email>(`provided email ${email} is not matched email pattern`)
    return Result.ok<Email>(new Email({ value: email }))
  }
}

```
So value object encapsulates some basic validation logic. The domain entities will look similar, but they encapsulate some more global business rules, for example, if we have user entity, it can ensure user type invariants (google or facebook or local user). Since domain entity construction can be complex, usually static factory method is used. 

So can I tell micro-orm that it should use some special method to create entity?
And if there is not such mechanism, can you give me advice about how I can deal with mapping in such case?

Sorry for my English, i am not native speaker :)

**Comment by B4nan** [comment, marked as answer]
It's up to you, you can either use decorators on your existing entities, or `EntitySchema` where you can pass the class you want to use.

https://mikro-orm.io/docs/defining-entities#classes-and-decorators
https://mikro-orm.io/docs/entity-schema#using-custom-entity-classes

For value objects you could use custom types or embeddables:

https://mikro-orm.io/docs/custom-types
https://mikro-orm.io/docs/embeddables

**Reply by B4nan** [reply]
That is pretty much what embeddables are for, you can disable prefixing and use the default inlined mode. But currently it does not support more than scalar properties, so no relations (I hope I will be able to relax this limitation at some point, for inlined mode it should be quite easy including native FKs).

You could try to implement custom hydrator, there are no docs for that, but it is configurable and relatively easy to implement. Hydrator is where the data loaded from database are mapped to entity properties.

https://mikro-orm.io/docs/configuration/#custom-hydrator

Current implementation uses JIT compiled functions, so it is a bit harder to read:

https://github.com/mikro-orm/mikro-orm/blob/master/packages/core/src/hydration/ObjectHydrator.ts

But there might be more issues with this approach, like change set computation might not work properly...

**Reply by MaKTD** [reply]
Yeah I read about custom types and embeddables, thanks to point to them, and now I think they will give me quite flexible way to organize mapping (that's why delete second answer). There will be some trade off, because i still need to place relations at the top level, but since it will only needed for entities and aggregates, I think it's worth it. Moreover maybe there will be some be advantages of sepparting entities from valueobjects inside aggregate roots. Anyway micro will dramatically simplify persistence and mapping logic for me, so I will definitely try.   


That would be amazing
`(I hope I will be able to relax this limitation at some point, for inlined mode it should be quite easy including native FKs )` 


thanks for hudrator point. I will return to it, if there will be difficulties in wiring everything up
